runners/samza/src/main/java/org/apache/beam/runners/samza/metrics/SamzaMetricsContainer.java

```diff
@@ -81,15 +106,133 @@ public MetricsContainerStepMap getContainers() {
     return this.metricsContainers;
   }
 
-  /** Update Beam metrics to Samza metrics for the current step and global step. */
+  /**
+   * This is the public method for updating metrics. It either defers the update to the executor
+   * service or updates the metrics immediately based on the deferToExecutor configuration flag.
+   *
+   * @param stepName the step name for which metrics are being updated
+   */
   public void updateMetrics(String stepName) {
+    if (deferToExecutor) {
```

Comment:
Why not start the thread pool when a SamzaMetricsContainer instance in created?

Reply:
What will happen if multiple threads call this method concurrently?

Reply:
After thinking about this more it's better to move this logic into constructor, which will remove the need for volatile and double check locking.